src/checkers/inference/solver/backend/SolverAdapter.java

```diff
@@ -0,0 +1,126 @@
+ * called in class {@link checkers.inference.solver.SolverEngine}}. See
+ * {@link checkers.inference.solver.SolverEngine#solveInparall()} and
+ * {@link checkers.inference.solver.SolverEngine#solveInSequential()}.
+ * 
+ * @author jianchu
+ *
+ * @param <T> type of FormatTranslator required by this SolverAdapter
+ */
+public abstract class SolverAdapter<T extends FormatTranslator<?, ?, ?>> {
+
+    /**
+     * String key value pairs to configure the solver
+     */
+    protected final Map<String, String> configuration;
+
+    /**
+     * Collection of all slots will be used by underlying solver
+     */
+    protected final Collection<Slot> slots;
+
+    /**
+     * Collection of all constraints will be solved by underlying solver
+     */
+    protected final Collection<Constraint> constraints;
+
+    /**
+     * Target QualifierHierarchy
+     */
+    protected final QualifierHierarchy qualHierarchy;
+
+    protected final ProcessingEnvironment processingEnvironment;
+
+    /**
+     * translator for encoding inference slots and constraints to underlying solver's constraints,
+     * and decoding underlying solver's solution back to AnnotationMirrors.
+     */
+    protected final T formatTranslator;
+
+    /**
+     * Set of ids of all variable solts will be used by underlying solver
+     */
+    protected final Set<Integer> varSlotIds;
+
+    /**
+     * Target qualifier lattice
+     */
+    protected final Lattice lattice;
+
+    public SolverAdapter(Map<String, String> configuration, Collection<Slot> slots,
+            Collection<Constraint> constraints, QualifierHierarchy qualHierarchy,
+            ProcessingEnvironment processingEnvironment, T formatTranslator, Lattice lattice) {
+        this.configuration = configuration;
+        this.slots = slots;
+        this.constraints = constraints;
+        this.qualHierarchy = qualHierarchy;
+        this.processingEnvironment = processingEnvironment;
+        this.formatTranslator = formatTranslator;
+        this.varSlotIds = new HashSet<Integer>();
+        this.lattice = lattice;
+    }
+
+    /**
+     * A concrete solver adapter needs to override this method and implements its own
+     * constraint-solving strategy. In general, there will be three steps in this method:
+     * 1. Calls {@link #convertAll()}, let {@link FormatTranslator} to convert constraints into
+     * the corresponding encoding form.
+     * 2. Calls the underlying solver to solve the encoding.
+     * 3. Let {@link FormatTranslator} decodes the solution from the underlying solver and create a map between an 
+     * Integer(Slot Id) and an AnnotationMirror as it's inferred annotation. 
+     * 
+     * It is the concrete solver adapter's responsibility to implemented the logic of above instructions and statistic collection. 
+     * See {@link checkers.inference.solver.backend.maxsat.MaxSatSolver#solve()}} for an example.
+     */
+    public abstract Map<Integer, AnnotationMirror> solve();
```

Comment:
It is better to refactor this class:

- Have an interface define the public API for `solve()`

- Extract common statistic/ other things into a default implementation of the interface

- Remove abstract methods that are not part of the template defined in default implementation